pkg/tecdsa/dkg/states.go

```diff
@@ -416,167 +301,93 @@ func (trts *tssRoundThreeState) MemberIndex() group.MemberIndex {
-
-	if len(fs.member.group.InactiveMemberIDs()) > 0 {
-		return newInactiveMembersError(fs.member.group.InactiveMemberIDs())
+	message, err := fs.member.tssFinalize(
+		ctx,
+		receivedMessages[*tssRoundThreeMessage](fs.BaseAsyncState),
+	)
+	if err != nil {
+		return err
 	}
 
-	// TSS computations can be time-consuming and can exceed the current
-	// state's time window. The ctx parameter is scoped to the lifetime of
-	// the current state so, it can be used as a timeout signal. However,
-	// that ctx is cancelled upon state's end only after Initiate returns.
-	// In order to make that working, Initiate must trigger the computations
-	// in a separate goroutine and return before the end of the state.
-	go func() {
-		message, err := fs.member.tssFinalize(ctx, fs.previousPhaseMessages)
-		if err != nil {
-			fs.outcomeChan <- err
-			return
-		}
-
-		if err := fs.channel.Send(ctx, message); err != nil {
-			fs.outcomeChan <- err
-			return
-		}
-
-		close(fs.outcomeChan)
-	}()
+	if err := fs.channel.Send(
+		ctx,
+		message,
+		net.BackoffRetransmissionStrategy,
+	); err != nil {
+		return err
+	}
 
 	return nil
 }
 
-func (fs *finalizationState) Receive(msg net.Message) error {
-	switch phaseMessage := msg.Payload().(type) {
-	case *tssFinalizationMessage:
+func (fs *finalizationState) Receive(netMessage net.Message) error {
+	if protocolMessage, ok := netMessage.Payload().(message); ok {
 		if fs.member.shouldAcceptMessage(
-			phaseMessage.SenderID(),
-			msg.SenderPublicKey(),
-		) && fs.member.sessionID == phaseMessage.sessionID {
-			fs.phaseMessages = append(fs.phaseMessages, phaseMessage)
+			protocolMessage.SenderID(),
+			netMessage.SenderPublicKey(),
+		) && fs.member.sessionID == protocolMessage.SessionID() {
+			fs.ReceiveToHistory(netMessage)
 		}
 	}
 
 	return nil
 }
 
-func (fs *finalizationState) Next() (state.SyncState, error) {
-	err := <-fs.outcomeChan
-	if err != nil {
-		return nil, err
-	}
-
-	return &confirmationState{
-		channel:               fs.channel,
-		member:                fs.member.initializeConfirmation(),
-		previousPhaseMessages: fs.phaseMessages,
-	}, nil
-}
-
-func (fs *finalizationState) MemberIndex() group.MemberIndex {
-	return fs.member.id
-}
-
-// confirmationState is the state during which members confirm a successful
-// course of the distributed key generation process.
-type confirmationState struct {
-	channel net.BroadcastChannel
-	member  *confirmingMember
-
-	previousPhaseMessages []*tssFinalizationMessage
-}
-
-func (cs *confirmationState) DelayBlocks() uint64 {
-	return silentStateDelayBlocks
-}
-
-func (cs *confirmationState) ActiveBlocks() uint64 {
-	return silentStateActiveBlocks
-}
-
-func (cs *confirmationState) Initiate(ctx context.Context) error {
-	cs.member.markInactiveMembers(cs.previousPhaseMessages)
-
-	if len(cs.member.group.InactiveMemberIDs()) > 0 {
-		return newInactiveMembersError(cs.member.group.InactiveMemberIDs())
-	}
+func (fs *finalizationState) CanTransition() bool {
```

Comment:
How do we ensure all members had a chance of receiving messages in this state? 

https://github.com/keep-network/keep-core/blob/254aa487f60fdc1662d10a77fd55c4a6ecb4d4d4/pkg/protocol/state/state.go#L92-L125

Reply:
Basically, if a given member collected all the required messages to exit that state, we can say for sure that:
- the given member sent their message (their `Initiate` function completed for sure)
- other members are listening for that message (if other members sent their messages, their `Initiate` function is completed for sure and they are listening for incoming messages)

With the current construction of the async machine and the DKG protocol, there is no option that other members are behind the finalization state when the given member exits. I think that should be enough to fulfill the requirement "all members had a chance of receiving messages in this state". Am I missing something?

Reply:
I am thinking about the following scenario:
1. We have two members that exited `tssRoundThreeState` more or less at the same time.
2. There was some delay with the second member and they entered `finalizationState` after the first member completed `Initiate` and sent their message.
3. Now the first member waits for a message from the second member and the second member waits for a retransmission of a message already sent by the first member.
4. The second member completes `Initiate`, and they sent their message.
5. The first member has all the messages and is exiting the state machine, closing the context, before the retransmission happens.
6. The second member hangs.

```
M1:   | Initiate() + send ↓                     | context.Done()         R <- here we would have a retransmission         
M2:                         | Initiate + send ↑ ..................
```

Reply:
Yeah, I got it. If we ask "did M2 have a chance to receive M1's message?" the answer is "yes, it had". M2 listened to M1's message from the very beginning and M1 sent it for sure and there is a chance it retransmitted it at least once. The real question is "did M2 have enough time to receive M1's message?". Well, this is not so simple and I'm not sure if we can guarantee that on the machine level. We can add a delay for M1, of course. But, what is the right duration of the delay? We can also extend the broadcast channel's `Send` by adding an option to guarantee N retransmissions. Not sure if this is the right direction though.

Reply:
After revisiting it one more time, I think we are actually good. The context for an attempt is active until the timeout for the attempt, even if the attempt finished earlier. It means that for a good timeout, we should have enough retransmissions for everyone:

https://github.com/keep-network/keep-core/blob/254aa487f60fdc1662d10a77fd55c4a6ecb4d4d4/pkg/tbtc/node.go#L534-L553

Does it make sense?

Reply:
Ah yeah! That's right!